File: tests/priority.c
/* Conformance test of the priority group on Linux: every answer is checked against
 * getpriority and against the stat files of procfs, for this process with several
 * threads, for a child by its id and for a child with several threads. */
#define _GNU_SOURCE
#include "dotnet_pal.h"
#include <assert.h>
#include <dirent.h>
#include <errno.h>
#include <grp.h>
#include <limits.h>
#include <pthread.h>
#include <signal.h>
#include <stdatomic.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/resource.h>
#include <sys/syscall.h>
#include <sys/wait.h>
#include <unistd.h>
#ifdef PAL_HOST_TEST
extern int pal_priority_fault;
#endif
#define INVALID DOTNET_PAL_INVALID_ARGUMENT
#define DENIED DOTNET_PAL_ACCESS_DENIED
#define MISSING DOTNET_PAL_NOT_FOUND
static const dotnet_pal_priority_ops *p;
static dotnet_pal_priority_stats expected;
static atomic_int stop, own_value;
static atomic_long idle_thread, other_thread;

/* One get: the status, and the value, which a refusal leaves 0. */
static uint32_t get(uint64_t process, int32_t *value) {
    *value = 77;
    uint32_t status = p->get(process, value);
    if (status == 0) ++expected.get_ok; else { ++expected.rejected_or_failed; assert(*value == 0); }
    return status;
}
static uint32_t set(uint64_t process, int32_t value) {
    uint32_t status = p->set(process, value);
    if (status == 0) ++expected.set_ok; else ++expected.rejected_or_failed;
    return status;
}
static int nice_of(id_t thread) { errno = 0; int value = getpriority(PRIO_PROCESS, thread); assert(value != -1 || errno == 0); return value; }
/* The nice value in a stat file of procfs: the 19th field, counted behind the command. */
static int stat_nice(const char *path) {
    char line[1024];
    FILE *file = fopen(path, "r");
    assert(file && fgets(line, sizeof line, file));
    fclose(file);
    char *at = strrchr(line, ')');
    for (int field = 2; field < 19; ++field) { assert(at); at = strchr(at + 1, ' '); }
    assert(at);
    return atoi(at + 1);
}
/* Every thread of the process has the value, by getpriority and by procfs; the number of threads. Callers ask for
 * at least the threads they started: a sanitizer runtime adds one of its own at a moment of its choosing. */
static int whole_process(pid_t process, int value) {
    char path[320]; int threads = 0; struct dirent *entry;
    snprintf(path, sizeof path, "/proc/%d/task", (int)process);
    DIR *list = opendir(path);
    assert(list);
    while ((entry = readdir(list)) != NULL) {
        if (entry->d_name[0] == '.') continue;
        snprintf(path, sizeof path, "/proc/%d/task/%s/stat", (int)process, entry->d_name);
        assert(stat_nice(path) == value && nice_of((id_t)atoi(entry->d_name)) == value);
        ++threads;
    }
    closedir(list);
    return threads;
}
static void *idle(void *argument) {
    atomic_store((atomic_long *)argument, syscall(SYS_gettid));
    while (!atomic_load(&stop)) usleep(1000);
    return NULL;
}
/* A thread that gives itself another value than its process has: to the kernel "0" is this thread, to the boundary it is this process. */
static void *apart(void *argument) {
    int32_t value; long thread = syscall(SYS_gettid);
    assert(setpriority(PRIO_PROCESS, (id_t)thread, 9) == 0 && nice_of(0) == 9 && nice_of((id_t)getpid()) == 5);
    uint32_t status = p->get(0, &value);
    assert(status == 0 && value == 5);
    atomic_store(&own_value, 1);
    return idle(argument);
}
/* Without privileges: the priority goes down and not up again, and another user's process is out of reach. */
static int unprivileged(pid_t parent) {
    int32_t value; dotnet_pal_priority_stats before, after, mine = expected;
    assert(p->read_stats(&before, sizeof before) == 0);
    if (setgroups(0, NULL) != 0 || setgid(54321) != 0 || setuid(54321) != 0) return 77;
    assert(set(0, 0) == DENIED && get(0, &value) == 0 && value == 10 && nice_of(0) == 10);
    assert(set(0, 9) == DENIED && set(0, 12) == 0 && get(0, &value) == 0 && value == 12 && nice_of(0) == 12);
    assert(set((uint64_t)parent, 15) == DENIED && get((uint64_t)parent, &value) == 0 && value == 10 && nice_of((id_t)parent) == 10);
    assert(p->read_stats(&after, sizeof after) == 0);
    assert(after.get_ok - before.get_ok == expected.get_ok - mine.get_ok && after.set_ok - before.set_ok == expected.set_ok - mine.set_ok);
    assert(after.rejected_or_failed - before.rejected_or_failed == expected.rejected_or_failed - mine.rejected_or_failed && after.rejected_or_failed - before.rejected_or_failed == 3);
    return 0;
}
/* A child with two more threads that tells through the pipe when they run, and then waits to be ended. */
static pid_t threaded_child(void) {
    int ready[2]; char byte;
    assert(pipe(ready) == 0);
    pid_t child = fork(); assert(child >= 0);
    if (child == 0) {
        pthread_t threads[2]; atomic_long ids[2] = {0, 0};
        for (int i = 0; i < 2; ++i) if (pthread_create(&threads[i], NULL, idle, &ids[i]) != 0) _exit(1);
        while (!atomic_load(&ids[0]) || !atomic_load(&ids[1])) usleep(1000);
        if (write(ready[1], "r", 1) != 1) _exit(1);
        for (;;) pause();
    }
    close(ready[1]);
    assert(read(ready[0], &byte, 1) == 1);
    close(ready[0]);
    return child;
}
/* Whether this process may make itself more favoured again: CAP_SYS_NICE, which the root of a default container has not got, or a nice limit that allows it. */
static int may_lower(void) {
    pid_t child = fork(); assert(child >= 0);
    if (child == 0) _exit(setpriority(PRIO_PROCESS, 0, 19) == 0 && setpriority(PRIO_PROCESS, 0, -20) == 0 ? 0 : 1);
    int result = 0;
    assert(waitpid(child, &result, 0) == child && WIFEXITED(result));
    return WEXITSTATUS(result) == 0;
}
static void reap(pid_t child) { int result; assert(kill(child, SIGKILL) == 0 && waitpid(child, &result, 0) == child); }

int main(int argc, char **argv) {
#ifdef PAL_HOST_TEST
    pal_priority_fault = argc > 1 ? atoi(argv[1]) : 0;
#else
    (void)argc; (void)argv;
#endif
    const dotnet_pal_api *api = dotnet_pal_get_api(2);
#ifdef PAL_HOST_TEST
    if (pal_priority_fault == 1 || pal_priority_fault == 3 || pal_priority_fault == 4) { assert(!api); puts("PRIORITY malformed host rejected"); return 0; }
#endif
    assert(api && api->header.struct_size >= DOTNET_PAL_PRIORITY_API_SIZE);
    assert(api->header.capabilities & DOTNET_PAL_CAP_PRIORITY);
    p = &api->priority;
    assert(p->get && p->set && p->read_stats);
    int32_t value; dotnet_pal_priority_stats stats;
#ifdef PAL_HOST_TEST
    if (pal_priority_fault == 2) {
        /* A value outside the range, an unknown status, a status get does not have: refused, and the value stays 0. */
        for (int i = 0; i < 7; ++i) assert(get(0, &value) == DOTNET_PAL_OS_ERROR);
        assert(get(0, &value) == MISSING); /* no such process is get's to report, without the output of the call that reported it */
        assert(get(0, &value) == 0 && value == 19 && get(0, &value) == 0 && value == -20);
        assert(set(0, 1) == DOTNET_PAL_OS_ERROR && set(0, 1) == DOTNET_PAL_OS_ERROR);
        /* A value that is none does not reach the provider: its next answer is still the third one. */
        assert(set(0, 20) == INVALID && set(0, 1) == DENIED && set(0, 1) == MISSING && set(0, 1) == 0);
        assert(p->read_stats(&stats, sizeof stats) == 0 && stats.get_ok == 2 && stats.set_ok == 1 && stats.rejected_or_failed == 8 + 5);
        assert(expected.get_ok == 2 && expected.set_ok == 1 && expected.rejected_or_failed == 13);
        puts("PRIORITY host errors sanitized"); return 0;
    }
#endif
    /* This process, which has a second thread from the start: its value is what getpriority says, and a new value is the value of every thread. */
    pthread_t threads[3];
    assert(pthread_create(&threads[0], NULL, idle, &idle_thread) == 0);
    while (!atomic_load(&idle_thread)) usleep(1000);
    int start = nice_of(0), root = geteuid() == 0;
    assert(get(0, &value) == 0 && value == start && (start <= 5 || root));
    assert(set(0, 5) == 0 && nice_of(0) == 5 && nice_of((id_t)getpid()) == 5 && get(0, &value) == 0 && value == 5 && get((uint64_t)getpid(), &value) == 0 && value == 5);
    assert(whole_process(getpid(), 5) >= 2);
    /* A thread with a value of its own does not change what the process answers, and is brought back by the next change. */
    assert(pthread_create(&threads[1], NULL, apart, &other_thread) == 0);
    while (!atomic_load(&other_thread)) usleep(1000);
    assert(atomic_load(&own_value) && nice_of((id_t)atomic_load(&other_thread)) == 9 && nice_of(0) == 5);
    expected.get_ok += 1; /* the one that thread asked */
    assert(get(0, &value) == 0 && value == 5);
    assert(set(0, 10) == 0 && whole_process(getpid(), 10) >= 3 && get(0, &value) == 0 && value == 10);
    /* A thread started now inherits the value. */
    atomic_long late = 0;
    assert(pthread_create(&threads[2], NULL, idle, &late) == 0);
    while (!atomic_load(&late)) usleep(1000);
    assert(whole_process(getpid(), 10) >= 4);

    /* Going back up is a privilege. A child that has given its privileges up has not got it; this process has it or not, and a refusal changes nothing. */
    int unprivileged_child = -1, lowering = may_lower();
    if (root) {
        pid_t child = fork(); assert(child >= 0);
        if (child == 0) _exit(unprivileged(getppid()));
        int result = 0;
        assert(waitpid(child, &result, 0) == child && WIFEXITED(result) && (WEXITSTATUS(result) == 0 || WEXITSTATUS(result) == 77));
        unprivileged_child = WEXITSTATUS(result) == 0;
    }
    if (unprivileged_child != 1) puts("PRIORITY note: not root, or the ids could not be changed: a refused change was not checked in a child without privileges");
    if (lowering) {
        assert(set(0, 0) == 0 && whole_process(getpid(), 0) >= 4 && get(0, &value) == 0 && value == 0);
        assert(set(0, -20) == 0 && whole_process(getpid(), -20) >= 4 && get(0, &value) == 0 && value == -20);
        /* -1 is a value, not a failure. */
        assert(set(0, -1) == 0 && whole_process(getpid(), -1) >= 4 && get(0, &value) == 0 && value == -1);
        assert(set(0, 19) == 0 && whole_process(getpid(), 19) >= 4 && get(0, &value) == 0 && value == 19);
        assert(set(0, 0) == 0 && whole_process(getpid(), 0) >= 4);
    } else {
        assert(set(0, 0) == DENIED && set(0, -1) == DENIED && whole_process(getpid(), 10) >= 4 && get(0, &value) == 0 && value == 10);
        puts("PRIORITY note: this process may not make itself more favoured (a container without CAP_SYS_NICE): the values below the one it has, -1 among them, were not checked");
    }
    /* A child by its id: getpriority and procfs see the value the boundary set, and the boundary reads what they see.
     * The values stand above the one this process has now, so the test needs no privilege for them: 7, 8 and 3 as root. */
    char path[64]; int base = nice_of(0);
    assert(base <= 10);
    pid_t child = fork(); assert(child >= 0);
    if (child == 0) { execlp("sleep", "sleep", "60", (char *)NULL); _exit(127); }
    assert(set((uint64_t)child, base + 7) == 0 && nice_of((id_t)child) == base + 7 && get((uint64_t)child, &value) == 0 && value == base + 7);
    snprintf(path, sizeof path, "/proc/%d/stat", (int)child);
    assert(stat_nice(path) == base + 7);
    assert(setpriority(PRIO_PROCESS, (id_t)child, base + 8) == 0 && get((uint64_t)child, &value) == 0 && value == base + 8);
    /* A child with threads: all of them, where setpriority alone changes the first. */
    pid_t crowd = threaded_child();
    assert(setpriority(PRIO_PROCESS, (id_t)crowd, base + 3) == 0 && nice_of((id_t)crowd) == base + 3);
    snprintf(path, sizeof path, "/proc/%d/task", (int)crowd);
    int first_only = 0; struct dirent *entry; DIR *list = opendir(path);
    assert(list);
    while ((entry = readdir(list)) != NULL) if (entry->d_name[0] != '.' && atoi(entry->d_name) != crowd) { assert(nice_of((id_t)atoi(entry->d_name)) == base); ++first_only; }
    closedir(list);
    assert(first_only == 2);
    assert(set((uint64_t)crowd, base + 8) == 0 && whole_process(crowd, base + 8) >= 3 && get((uint64_t)crowd, &value) == 0 && value == base + 8);
    reap(crowd);
    /* A process that has ended and been waited for is no process. */
    reap(child);
    assert(get((uint64_t)child, &value) == MISSING && set((uint64_t)child, 7) == MISSING);
    assert(get((uint64_t)crowd, &value) == MISSING && set((uint64_t)crowd, 7) == MISSING);

    /* An id no process can have is no process, and never this one or another by the low half of the number. */
    uint64_t none[] = {(uint64_t)INT32_MAX + 1, UINT32_MAX, (UINT64_C(1) << 32), (UINT64_C(1) << 32) + (uint64_t)getpid(), (UINT64_C(1) << 63) + (uint64_t)getpid(), UINT64_MAX};
    for (size_t i = 0; i < sizeof none / sizeof *none; ++i) assert(get(none[i], &value) == MISSING && set(none[i], 15) == MISSING);
    assert(nice_of(0) == base && get(0, &value) == 0 && value == base);
    /* Values that are none, and an answer that has no place. */
    int32_t wrong[] = {-21, 20, INT32_MIN, INT32_MAX};
    for (size_t i = 0; i < sizeof wrong / sizeof *wrong; ++i) assert(set(0, wrong[i]) == INVALID && set((uint64_t)getpid(), wrong[i]) == INVALID);
    assert(p->get(0, NULL) == INVALID && p->get(0, (int32_t *)((uint8_t *)&value + 1)) == INVALID);
    expected.rejected_or_failed += 2;
    assert(p->read_stats(NULL, sizeof stats) == INVALID && p->read_stats(&stats, sizeof stats - 1) == INVALID);

    atomic_store(&stop, 1);
    for (int i = 0; i < 3; ++i) assert(pthread_join(threads[i], NULL) == 0);
    /* Counters: one for the values read, one for the values set, one for everything refused. */
    assert(p->read_stats(&stats, sizeof stats) == 0);
    assert(stats.get_ok == expected.get_ok && stats.set_ok == expected.set_ok && stats.rejected_or_failed == expected.rejected_or_failed);
    printf("PRIORITY PASS start=%d root=%d unprivileged_child=%d lowering=%d get_ok=%llu set_ok=%llu refused=%llu\n", start, root, unprivileged_child, lowering,
        (unsigned long long)stats.get_ok, (unsigned long long)stats.set_ok, (unsigned long long)stats.rejected_or_failed);
    return 0;
}

File: crates/dotnet-pal-std/tests/priority.rs
//! Exercises the priority group of the std port through the negotiated C table, on
//! whatever Unix runs the test, against getpriority and, on Linux, the stat files of
//! procfs. What changes the priority of the test itself runs in a child of its own.
#![cfg(unix)]
use dotnet_pal_rs::io::ACCESS_DENIED;
use dotnet_pal_rs::priority::{self, Stats};
use dotnet_pal_rs::runtime::NOT_FOUND;
use dotnet_pal_rs::{INVALID_ARGUMENT, OK};
use std::process::{Child, Command};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Mutex, MutexGuard};
use std::{mem::size_of, ptr};

const CHILD: &str = "PAL_STD_PRIORITY_CHILD";
/// One test at a time: the counters are the process's.
static SERIAL: Mutex<()> = Mutex::new(());
fn serial() -> MutexGuard<'static, ()> { SERIAL.lock().unwrap_or_else(|poisoned| poisoned.into_inner()) }
fn ops() -> &'static priority::Ops {
    let api = dotnet_pal_std::api();
    assert!(!api.is_null());
    let api = unsafe { &*api };
    assert_eq!(api.header.capabilities & priority::CAP, priority::CAP);
    &api.priority
}
/// What the calls of this test should have added to the three counters.
static TALLY: [AtomicU64; 3] = [const { AtomicU64::new(0) }; 3];
fn tally(status: u32, index: usize) -> u32 { TALLY[if status == OK { index } else { 2 }].fetch_add(1, Ordering::Relaxed); status }
fn counters() -> ([u64; 3], [u64; 3]) {
    let mut out = Stats::default();
    assert_eq!(unsafe { ops().read_stats.unwrap()(&mut out, size_of::<Stats>()) }, OK);
    ([out.get_ok, out.set_ok, out.rejected_or_failed], [0, 1, 2].map(|index| TALLY[index].load(Ordering::Relaxed)))
}
/// Runs `body` and checks that the counters moved by exactly what its calls should have added.
fn counted(body: impl FnOnce()) {
    let (stats, tally) = counters();
    body();
    let (stats_after, tally_after) = counters();
    assert_eq!([0, 1, 2].map(|index| stats_after[index] - stats[index]), [0, 1, 2].map(|index| tally_after[index] - tally[index]));
}
/// Status and value of one get; the value starts as garbage.
fn get(process: u64) -> (u32, i32) {
    let mut value = 77;
    (tally(unsafe { ops().get.unwrap()(process, &mut value) }, 0), value)
}
fn set(process: u64, value: i32) -> u32 { tally(unsafe { ops().set.unwrap()(process, value) }, 1) }
/// What getpriority says about a process or, on Linux, a thread. -1 is a value: only errno tells it from a failure.
fn nice_of(id: u32) -> i32 {
    #[cfg(target_os = "linux")]
    unsafe { *libc::__errno_location() = 0 };
    #[cfg(target_os = "macos")]
    unsafe { *libc::__error() = 0 };
    let value = unsafe { libc::getpriority(libc::PRIO_PROCESS, id as libc::id_t) };
    assert!(value != -1 || std::io::Error::last_os_error().raw_os_error().unwrap_or(0) == 0);
    value
}
/// The nice values procfs shows for every thread of a process.
#[cfg(target_os = "linux")]
fn thread_values(process: u32) -> Vec<i32> {
    std::fs::read_dir(format!("/proc/{process}/task")).unwrap().map(|thread| {
        let stat = std::fs::read_to_string(thread.unwrap().path().join("stat")).unwrap();
        // The 19th field, counted behind the command, which may hold spaces and parentheses.
        stat[stat.rfind(')').unwrap() + 2..].split(' ').nth(16).unwrap().parse().unwrap()
    }).collect()
}
/// Whether a process like this one may make itself more favoured again: root outside a container, or a nice limit that allows it.
fn may_lower() -> bool {
    match unsafe { libc::fork() } {
        -1 => panic!("fork"),
        // Only calls a forked child of a threaded process may make.
        0 => unsafe { libc::_exit(if libc::setpriority(libc::PRIO_PROCESS, 0, 19) == 0 && libc::setpriority(libc::PRIO_PROCESS, 0, -20) == 0 { 0 } else { 1 }) },
        child => { let mut status = 0; assert_eq!(unsafe { libc::waitpid(child, &mut status, 0) }, child); libc::WIFEXITED(status) && libc::WEXITSTATUS(status) == 0 }
    }
}
struct Reaped(Child);
impl Drop for Reaped { fn drop(&mut self) { let _ = self.0.kill(); let _ = self.0.wait(); } }
fn this_test(mode: &str) -> Command {
    let mut command = Command::new(std::env::current_exe().unwrap());
    command.args(["--exact", "child_mode", "--nocapture", "--test-threads=1"]).env(CHILD, mode);
    command
}

#[test]
fn this_process_by_zero_and_by_its_id() {
    let _serial = serial();
    let output = this_test("self").output().unwrap();
    assert!(output.status.success(), "{}{}", String::from_utf8_lossy(&output.stdout), String::from_utf8_lossy(&output.stderr));
    // The harness prints the name of the child's test in front of the line.
    for line in String::from_utf8_lossy(&output.stdout).lines() { if let Some(at) = line.find("priority:") { println!("{}", &line[at..]); } }
}
#[test]
fn child_mode() {
    let Some(mode) = std::env::var_os(CHILD) else { return; };
    if mode == "crowd" {
        // Two more threads, a line that says they run, and a wait to be ended.
        let barrier = std::sync::Arc::new(std::sync::Barrier::new(3));
        for _ in 0..2 { let barrier = barrier.clone(); std::thread::spawn(move || { barrier.wait(); loop { std::thread::park(); } }); }
        barrier.wait();
        println!("ready");
        loop { std::thread::park(); }
    }
    counted(|| {
        let (own, start) = (std::process::id(), nice_of(0));
        assert!(start <= 4, "the test starts from a nice value of 4 or less");
        assert_eq!(get(0), (OK, start));
        assert_eq!(get(own as u64), (OK, start));
        // The harness runs this test on a thread of its own, so the process has two threads at least: on Linux every one of them follows.
        assert_eq!(set(0, 5), OK);
        assert_eq!((get(0), nice_of(0), nice_of(own)), ((OK, 5), 5, 5));
        #[cfg(target_os = "linux")]
        {
            let values = thread_values(own);
            assert!(values.len() >= 2 && values.iter().all(|value| *value == 5), "{values:?}");
            // A thread with a value of its own does not change what the process answers, and is brought back by the next change.
            let (asked, answer) = (std::sync::Arc::new(std::sync::Barrier::new(2)), std::sync::Arc::new(std::sync::Barrier::new(2)));
            let apart = { let (asked, answer) = (asked.clone(), answer.clone()); std::thread::spawn(move || {
                let thread = unsafe { libc::syscall(libc::SYS_gettid) } as u32;
                assert_eq!(unsafe { libc::setpriority(libc::PRIO_PROCESS, thread as libc::id_t, 9) }, 0);
                assert_eq!((nice_of(0), get(0)), (9, (OK, 5)));
                asked.wait();
                answer.wait();
                assert_eq!(nice_of(0), 10);
            }) };
            asked.wait();
            assert!(thread_values(own).contains(&9));
            assert_eq!(set(own as u64, 10), OK);
            let values = thread_values(own);
            assert!(values.len() >= 3 && values.iter().all(|value| *value == 10), "{values:?}");
            answer.wait();
            apart.join().unwrap();
        }
        #[cfg(not(target_os = "linux"))]
        assert_eq!((set(own as u64, 10), get(0), nice_of(0)), (OK, (OK, 10), 10));
        // Going back up is a privilege, and a refusal changes nothing.
        let lowering = may_lower();
        if lowering {
            for value in [0, -20, -1, 19, 0] { assert_eq!((set(0, value), get(0), nice_of(0)), (OK, (OK, value), value)); }
        } else {
            assert_eq!((set(0, 0), set(0, 9), set(0, -1)), (ACCESS_DENIED, ACCESS_DENIED, ACCESS_DENIED));
            assert_eq!((get(0), nice_of(0)), ((OK, 10), 10));
            #[cfg(target_os = "linux")]
            assert!(thread_values(own).iter().all(|value| *value == 10));
        }
        println!("priority: start={start} lowering={lowering}");
    });
}
#[test]
fn a_child_by_its_id_until_it_is_gone() {
    let _serial = serial();
    counted(|| {
        let base = nice_of(0);
        assert!(base <= 10);
        let child = Reaped(Command::new("sleep").arg("60").spawn().unwrap());
        let id = child.0.id();
        assert_eq!(set(id as u64, base + 7), OK);
        assert_eq!((get(id as u64), nice_of(id)), ((OK, base + 7), base + 7));
        #[cfg(target_os = "linux")]
        assert_eq!(thread_values(id), [base + 7]);
        // What another program sets is what the boundary reads. macOS has a value 20, which is the boundary's 19.
        assert_eq!(unsafe { libc::setpriority(libc::PRIO_PROCESS, id as libc::id_t, base + 8) }, 0);
        assert_eq!(get(id as u64), (OK, base + 8));
        assert_eq!(unsafe { libc::setpriority(libc::PRIO_PROCESS, id as libc::id_t, 20) }, 0);
        assert!(nice_of(id) == if cfg!(target_os = "macos") { 20 } else { 19 });
        assert_eq!(get(id as u64), (OK, 19));
        // A process that has ended and been waited for is no process.
        drop(child);
        assert_eq!((get(id as u64), set(id as u64, 19)), ((NOT_FOUND, 0), NOT_FOUND));
    });
}
#[cfg(target_os = "linux")]
#[test]
fn every_thread_of_another_process() {
    use std::io::{BufRead, BufReader};
    let _serial = serial();
    counted(|| {
        let base = nice_of(0);
        let mut crowd = Reaped(this_test("crowd").stdout(std::process::Stdio::piped()).spawn().unwrap());
        let mut line = String::new();
        // The harness prints its own lines first, and the name of the child's test in front of the line.
        let mut output = BufReader::new(crowd.0.stdout.take().unwrap());
        while !line.trim_end().ends_with("ready") { line.clear(); assert!(output.read_line(&mut line).unwrap() > 0); }
        let id = crowd.0.id();
        let threads = thread_values(id).len();
        assert!(threads >= 3);
        // setpriority alone changes the first thread.
        assert_eq!(unsafe { libc::setpriority(libc::PRIO_PROCESS, id as libc::id_t, base + 3) }, 0);
        assert_eq!(thread_values(id).iter().filter(|value| **value == base + 3).count(), 1);
        assert_eq!(set(id as u64, base + 8), OK);
        assert_eq!((thread_values(id), get(id as u64)), (vec![base + 8; threads], (OK, base + 8)));
    });
}
#[test]
fn ids_and_values_that_are_none() {
    let _serial = serial();
    counted(|| {
        let (own, before) = (std::process::id() as u64, nice_of(0));
        // An id no process can have is no process, and never this one by the low half of the number.
        for process in [i32::MAX as u64 + 1, u32::MAX as u64, 1 << 32, (1 << 32) + own, (1 << 63) + own, u64::MAX] {
            assert_eq!((get(process), set(process, 15)), ((NOT_FOUND, 0), NOT_FOUND));
        }
        for value in [-21, 20, i32::MIN, i32::MAX] { assert_eq!((set(0, value), set(own, value)), (INVALID_ARGUMENT, INVALID_ARGUMENT)); }
        assert_eq!(nice_of(0), before);
        assert_eq!(tally(unsafe { ops().get.unwrap()(0, ptr::null_mut()) }, 0), INVALID_ARGUMENT);
        assert_eq!(unsafe { ops().read_stats.unwrap()(ptr::null_mut(), size_of::<Stats>()) }, INVALID_ARGUMENT);
        assert_eq!(unsafe { ops().read_stats.unwrap()(&mut Stats::default(), size_of::<Stats>() - 1) }, INVALID_ARGUMENT);
    });
}
